File: Java/app/src/main/java/dev/ogabek/java/adapter/FeedAdapter.java
package dev.ogabek.java.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.imageview.ShapeableImageView;
import com.tbuonomo.viewpagerdotsindicator.WormDotsIndicator;

import java.util.ArrayList;

import dev.ogabek.java.R;
import dev.ogabek.java.model.Feed;
import dev.ogabek.java.model.Post;

public class FeedAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final Context context;
    private final ArrayList<Feed> feeds;

    private static final int TYPE_STORY_VIEW = 0;
    private static final int TYPE_SIMPLE_POST_VIEW = 1;
    private static final int TYPE_PHOTOS_POST_VIEW = 2;

    public FeedAdapter(Context context, ArrayList<Feed> feeds) {
        this.context = context;
        this.feeds = feeds;
    }

    @Override
    public int getItemViewType(int position) {
        if (feeds.get(position).getStories().size() > 0) {
            return TYPE_STORY_VIEW;
        } else if (feeds.get(position).getPost().getPhotos().size() > 0) {
            return TYPE_PHOTOS_POST_VIEW;
        } else {
            return TYPE_SIMPLE_POST_VIEW;
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == TYPE_STORY_VIEW) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_feed_story, parent, false);
            return new StoryViewHolder(context, view);
        } else if (viewType == TYPE_PHOTOS_POST_VIEW) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_feed_post_photos, parent, false);
            return new PhotosPostViewHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_feed_post, parent, false);
            return new SimplePostViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Feed feed = feeds.get(position);

        if (holder instanceof StoryViewHolder) {
            ((StoryViewHolder) holder).recyclerView.setAdapter(new StoryAdapter(context, feed.getStories()));
        }

        if (holder instanceof SimplePostViewHolder) {
            Post post = feed.getPost();

            ((SimplePostViewHolder) holder).iv_profile.setImageResource(post.getProfile());
            ((SimplePostViewHolder) holder).iv_photo.setImageResource(post.getPhoto());
            ((SimplePostViewHolder) holder).tv_fullName.setText(post.getFullName());

        }

        if (holder instanceof PhotosPostViewHolder) {
            Post post = feed.getPost();

            ((PhotosPostViewHolder) holder).iv_profile.setImageResource(post.getProfile());
            ViewPagerAdapter adapter = new ViewPagerAdapter(context, feed.getPost().getPhotos());
            ((PhotosPostViewHolder) holder).vp_photo.setAdapter(adapter);
            ((PhotosPostViewHolder) holder).wormDotsIndicator.setViewPager(((PhotosPostViewHolder) holder).vp_photo);
            ((PhotosPostViewHolder) holder).tv_fullName.setText(post.getFullName());
        }

    }

    @Override
    public int getItemCount() {
        return feeds.size();
    }

    private static class StoryViewHolder extends RecyclerView.ViewHolder {

        RecyclerView recyclerView;

        public StoryViewHolder(Context context, View view) {
            super(view);

            recyclerView = view.findViewById(R.id.recyclerView);
            recyclerView.setLayoutManager(new GridLayoutManager(context, 1, LinearLayoutManager.HORIZONTAL, false));

        }
    }

    private static class SimplePostViewHolder extends RecyclerView.ViewHolder {

        ShapeableImageView iv_profile;
        TextView tv_fullName;
        ImageView iv_photo;

        public SimplePostViewHolder(View view) {
            super(view);
            iv_profile = view.findViewById(R.id.iv_profile);
            tv_fullName = view.findViewById(R.id.tv_full_name);
            iv_photo = view.findViewById(R.id.iv_photo);

        }
    }

    private static class PhotosPostViewHolder extends RecyclerView.ViewHolder {

        ShapeableImageView iv_profile;
        TextView tv_fullName;
        ViewPager vp_photo;
        WormDotsIndicator wormDotsIndicator;

        public PhotosPostViewHolder(View view) {
            super(view);

            iv_profile = view.findViewById(R.id.iv_profile);
            tv_fullName = view.findViewById(R.id.tv_full_name);
            vp_photo = view.findViewById(R.id.view_pager);
            wormDotsIndicator = view.findViewById(R.id.worm_dot);

        }
    }
}

File: Kotlin/app/src/main/java/dev/ogabek/kotlin/adapter/FeedAdapter.kt
package dev.ogabek.kotlin.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import com.google.android.material.imageview.ShapeableImageView
import com.tbuonomo.viewpagerdotsindicator.WormDotsIndicator
import dev.ogabek.kotlin.R
import dev.ogabek.kotlin.model.Feed
import dev.ogabek.kotlin.model.Post

class FeedAdapter(private val context: Context, private val feeds: ArrayList<Feed>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun getItemViewType(position: Int): Int {
        return when {
            feeds[position].stories.size > 0 -> {
                TYPE_STORY_VIEW
            }
            feeds[position].post!!.photos.isNotEmpty() -> {
                TYPE_PHOTOS_POST_VIEW
            }
            else -> {
                TYPE_SIMPLE_POST_VIEW
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == TYPE_STORY_VIEW) {
            val view =
                LayoutInflater.from(parent.context).inflate(R.layout.item_feed_story, parent, false)
            StoryViewHolder(context, view)
        } else if (viewType == FeedAdapter.TYPE_PHOTOS_POST_VIEW) {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_feed_post_photos, parent, false)
            PhotosPostViewHolder(view)
        } else {
            val view =
                LayoutInflater.from(parent.context).inflate(R.layout.item_feed_post, parent, false)
            SimplePostViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val feed: Feed = feeds[position]
        if (holder is StoryViewHolder) {
            holder.recyclerView.adapter =
                StoryAdapter(context, feed.stories)
        }
        if (holder is SimplePostViewHolder) {
            val post: Post = feed.post!!
            holder.iv_profile.setImageResource(post.profile)
            holder.iv_photo.setImageResource(post.photo)
            holder.tv_fullName.setText(post.fullName)
        }

        if (holder is PhotosPostViewHolder) {
            val post: Post = feed.post!!
            holder.iv_profile.setImageResource(post.profile)
            val adapter = ViewPagerAdapter(context, post.photos)
            holder.vp_photo.adapter = adapter
            holder.wormDotsIndicator.setViewPager(holder.vp_photo)
            holder.tv_fullName.setText(post.fullName)
        }
    }

    override fun getItemCount(): Int {
        return feeds.size
    }

    private class StoryViewHolder(context: Context?, view: View) :
        RecyclerView.ViewHolder(view) {
        var recyclerView: RecyclerView = view.findViewById(R.id.recyclerView)

        init {
            recyclerView.layoutManager =
                GridLayoutManager(context, 1, LinearLayoutManager.HORIZONTAL, false)
        }
    }

    private class SimplePostViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var iv_profile: ShapeableImageView = view.findViewById(R.id.iv_profile)
        var tv_fullName: TextView = view.findViewById(R.id.tv_full_name)
        var iv_photo: ImageView = view.findViewById(R.id.iv_photo)

    }

    private class PhotosPostViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var iv_profile: ShapeableImageView = view.findViewById(R.id.iv_profile)
        var tv_fullName: TextView = view.findViewById(R.id.tv_full_name)
        var vp_photo: ViewPager = view.findViewById(R.id.view_pager)
        var wormDotsIndicator: WormDotsIndicator = view.findViewById(R.id.worm_dot)

    }

    companion object {
        private const val TYPE_STORY_VIEW = 0
        private const val TYPE_SIMPLE_POST_VIEW = 1
        private const val TYPE_PHOTOS_POST_VIEW = 2
    }

}